src/key.rs

```diff
@@ -615,24 +622,29 @@ impl From<ffi::PublicKey> for PublicKey {
 #[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
 impl ::serde::Serialize for PublicKey {
     fn serialize<S: ::serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        // FIXME: human_readable serializes the whole 64 bytes but non-human-readable does 33 (compressed). Is this correct?
         if s.is_human_readable() {
             s.collect_str(self)
         } else {
-            s.serialize_bytes(&self.serialize())
+            let mut seq = s.serialize_tuple(constants::PUBLIC_KEY_SIZE)?;
+            for byte in self.serialize().iter() { // Serialize in compressed form.
+                seq.serialize_element(&byte)?;
+            }
+            seq.end()
         }
     }
 }
 
-#[cfg(feature = "serde")]
-#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+#[cfg(all(feature = "serde", any(feature = "std", feature = "alloc")))]
+#[cfg_attr(docsrs, doc(cfg(all(feature = "serde", any(feature = "std", feature = "alloc")))))]
 impl<'de> ::serde::Deserialize<'de> for PublicKey {
     fn deserialize<D: ::serde::Deserializer<'de>>(d: D) -> Result<PublicKey, D::Error> {
         if d.is_human_readable() {
             d.deserialize_str(super::serde_util::FromStrVisitor::new(
                 "an ASCII hex string representing a public key"
             ))
         } else {
-            d.deserialize_bytes(super::serde_util::BytesVisitor::new(
+            d.deserialize_seq(super::serde_util::SeqVisitor::new(
```

Comment:
This should be `deserialize_tuple`

Reply:
Oh so implement `de::Visitor` for each fixed size type we have and use `deserialize_tuple`, interesting. I'll give it a shot.

Reply:
Yes, if you want to keep it generic I suspect you'll have to use `PhantomData<fn() -> T>`

Reply:
With a macro for the serde impls, and serializing with tuple this PR is way cleaner now!